File: go/parsers/interrupt.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Interrupt struct {
	gen *gen.Interrupt
}

func NewInterruptFields(g *gen.Interrupt) messages.InterruptFields {
	return &Interrupt{gen: g}
}

func (i *Interrupt) RequestID() uint64 {
	return i.gen.RequestID()
}

func (i *Interrupt) Options() map[string]any {
	return map[string]any{}
}

func InterruptToCapnproto(m *messages.Interrupt) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	interrupt, err := gen.NewRootInterrupt(seg)
	if err != nil {
		return nil, err
	}

	interrupt.SetRequestID(m.RequestID())

	data, err := msg.Marshal()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeInterrupt, data, nil), nil
}

func CapnprotoToInterrupt(data []byte) (*messages.Interrupt, error) {
	msg, err := capnp.Unmarshal(data)
	if err != nil {
		return nil, err
	}

	interrupt, err := gen.ReadRootInterrupt(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewInterruptWithFields(NewInterruptFields(&interrupt)), nil
}

File: python/wamp_msgs_capnp/parsers/interrupt.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.interrupt import Interrupt, IInterruptFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "interrupt.capnp")
interrupt_capnp = capnp.load(str(module_file))


class InterruptFields(IInterruptFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def options(self) -> dict[str, Any]:
        return {}


def interrupt_to_capnproto(i: Interrupt) -> bytes:
    interrupt = interrupt_capnp.Interrupt.new_message()
    interrupt.requestID = i.request_id

    packed_data = interrupt.to_bytes_packed()

    return helpers.prepend_header(Interrupt.TYPE, packed_data, b"")


def capnproto_to_interrupt(data: bytes) -> Interrupt:
    message_data, _ = helpers.extract_message(data)
    interrupt_obj = interrupt_capnp.Interrupt.from_bytes_packed(message_data)

    return Interrupt(InterruptFields(interrupt_obj))
